sil.abnf

```diff
@@ -0,0 +1,71 @@
+; Formal grammar definition for SIL (SILE Input Language) files
+;
+; Based on RFC 5234 (Augmented BNF for Syntax Specifications: ABNF)
+; Uses RFC 7405 (Case-Sensitive String Support in ABNF)
+
+; A master document can only have one top level content item, but we allow
+; loading of fragments as well which can have any number of top level content
+; items, hence valid grammar can be any number of content items.
+document = *content
+
+; Top level content can be any sequence of these things
+content =  environment
+content =/ comment
+content =/ text
+content =/ braced_content
+content =/ command
+
+; Environments come in two flavors, raw passthrough and regular. The difference
+; is what is allowed to terminate them and what escapes are needed for the
+; content in the middle.
+environment =  %s"\begin" [ options ] "{" command-id "}" content %s"\end{" command-id "}"
+environment =/ %s"\begin" [ options ] "{" passthrough-command-id "}" passthrough-content %s"\end{" passthrough-command-id "}"
+
+; Nothing to see here.
+comment = "%" utf8-octets CRLF
+
+; Input strings that are not wrapped in some SILE command.
+text = *text-char
+
+; Input content wrapped in braces can be attatched to a command or just used to
+; create a settings wrapper.
```

Comment:
I'm not sure what a "settings wrapper" is.

The name might imply it does a temporary settings context, which is not true:

```
\begin{document}

Before {\set[parameter=font.size, value=30]Inside} Outside

\end{document}
```

![image](https://user-images.githubusercontent.com/18075640/222031385-2abdad9b-5776-4a7a-9428-cadb7adbfbc0.png)

Reply:
I believe that was the intended use, the fact that it is not working (stopped working?) is a bug. We did fix the `\set` command so that it can be used as a wrapper itself which was broken for a long time, but it seems to me what you show is a regression (unless I understand something wrong).

```sile
Before \set[parameter=font.size, value=30]{Inside} Outside
```

Reply:
I can think to one usage where grouping with braces was probably _not_ expected to result in anything else than AST grouping without temporary settings: the **gutenberg** package, which uses `\alternative{{and}{&}}`.

But it seems to be some weird attempt at working around the "one argument" only in SILE commands, and actually manage more than one argument under the hood...

It _could_ be interested with that purpose in mind, but with also XML parity in mind, I then wonder how to make sense of it.. I mean, if `\command[options][content}` more or less equates to `<command options>content</command>`, I am not sure to understand how the above "alternative" construct would translate. An implicit tag? Huh.

Reply:
Yes lets think about XML parity and how other formats might parse and generate a compatible AST. There is no sense in cramming some language feature in SIL markup that is hard/impossible in other input formats. I can see how it current gives deliberate meaning to `content[1]`, `content[2]`, etc. in the AST and I have no idea how to do that in XML.

Reply:
I've played around with this a bit and it doesn't work the way I would have wanted it to, but since the thing I wast also makes compatibility with other inputters tricky, I think I'll forgo fixing/implementing what I expected here for now.

In the mean time I think we should make `\set{}` work as a settings wrapper even with no settings in the actual command.